src/sagemaker_containers/transformer.py

```diff
@@ -0,0 +1,162 @@
+
+        The input_fn is responsible to take the request data and pre-process it before prediction.
+
+    Args:
+        input_data (obj): the request data content.
+        content_type (str): the request Content-Type.
+
+    Returns:
+        (obj): data ready for prediction.
+    """
+    return encoders.default_decoder.decode(input_data, content_type)
+
+
+def default_predict_fn(model, data):
+    """Function responsible for model predictions.
+
+    Args:
+        model (obj): model loaded by model_fn
+        data: de-serializes data returned by input_fn
+
+     Returns:
+         (obj): data ready for prediction.
+    """
+    raise NotImplementedError()
+
+
+def default_output_fn(prediction, accept):
+    """Function responsible to serialize the prediction for the response.
+
+    Args:
+        prediction (obj): prediction returned by predict_fn .
+        accept (str): accept content-type expected by the client.
+
+    Returns:
+        (worker.Response): a Flask response object with the following args:
+
+            * Args:
+                response: the serialized data to return
+                accept: the content-type that the data was transformed to.
+    """
+    return worker.Response(encoders.default_encoder.encode(prediction, accept), accept)
+
+
+class ClientError(BaseException):
+    pass
+
+
+class Transformer(object):
+    """The Transformer is a proxy between the worker and the framework transformation functions.
+
+    It implements the default framework functions for serving.
+
+    Examples:
+    >>>import os
+    >>>from sagemaker_containers import env, modules, transformer
+    >>>import Keras
+    >>>serving_env = env.ServingEnv()
+    >>>
+    >>>def predict_fn(model, data):
+    >>>     return model.predict(data)
+    >>>
+    >>>def model_fn(model_dir):
+    >>>     return Keras.models.load_model(os.path.join(model_dir, 'minimlmodel'))
+    >>>
+    >>>transformer = transformer.Transformer(predict_fn=predict_fn, model_fn=model_fn)
+    >>>
+    >>>mod = modules.download_and_import(serving_env.module_dir, serving_env.module_name)
+    >>>transformer.load_user_fns(mod)
+    """
+
+    def __init__(self, model_fn=None, input_fn=None, predict_fn=None, output_fn=None, error_class=ClientError):
+        """Default constructor. Wraps the any non default framework function in an error class to isolate
+        framework from user errors.
+
+        Args:
+            model_fn (fn): Function responsible to load the model.
+            input_fn (fn): Takes request data and de-serializes the data into an object for prediction.
+            predict_fn (fn): Function responsible for model predictions.
+            output_fn (fn): Function responsible to serialize the prediction for the response.
+            error_class (Exception): Error class used to separate framework and user errors.
+        """
+        self._model = None
+        self._model_fn = functions.error_wrapper(model_fn, error_class) or default_model_fn
```

Comment:
This line is unintuitive since I wouldn't expect error_wrapper to return None if model_fn is None. I think it'd be bettter like:

functions.error_wrapper(model_fn, error_class) if model_fn else default_model_fn

Reply:
Yeah, I agree.